### PR TITLE
Allow faster semantic classification for multiple text spans

### PR DESCRIPTION
**Pull request by maryamariyan**
### Description

Currently roslyn allows computing semantic classification by passing down a single text span for each document.
But for razor documents we'd like to allow requesting roslyn to get semantic classification by sending it multiple text spans within the document, without having it recompute metadata (mainly semantic model) per text span.

This PR adds a fast route for razor, by allowing roslyn server to compute semantic classification using multiple text spans.

### [before/after] 
- Commit https://github.com/dotnet/roslyn/commit/08c65b92b6841fa11967a9599577f97edfbe05cf shows code used to compute this data (with **caching disabled**)

image

<details>
 <summary> Versus when caching is enabled (DEFAULT) (click to view) </summary>

### [before/after] when caching is enabled 
- Commit https://github.com/dotnet/roslyn/commit/7c5f11365af756a64a1b62ab4d278b50960c5a9a shows code used to compute this data
 
image
</details>

<details>
 <summary> nullability analysis investigation (click to view) </summary>

For more information: https://github.com/dotnet/roslyn/issues/70609

### [before] when nullability analysis is disabled
- (without PR diff / `UsePreciseSemanticTokensRanges` feature flag OFF)
- Commit https://github.com/dotnet/roslyn/commit/a77837133b92a0e03f4077c6f89e59f52ab9152e shows code used to compute this data 
- To toggle nullability analysis, added PropertyGroup <Features>run-nullable-analysis=always</Features> to the file's csproj.

[a] - caching enabled / `run-nullable-analysis=always` **(DEFAULT BEHAVIOR TODAY)**
image
[b] - caching disabled / `run-nullable-analysis=never`
image
[c] - caching enabled / `run-nullable-analysis=never`
image

### [after] when nullability analysis is disabled
- (with PR diff / `UsePreciseSemanticTokensRanges` feature flag ON)
- Commit https://github.com/dotnet/roslyn/commit/ecf6c4e5de0e064302ee5cc361197745c1fe2753 shows code used to compute this data
[d] - caching disabled / `run-nullable-analysis=never`
image
[e] - caching enabled / `run-nullable-analysis=never`
image
[f] - caching enabled / `run-nullable-analysis=always` **(DEFAULT BEHAVIOR FOR CURRENT PR)**
image

</details>

### Repro steps for these screenshots
- Waited a few minutes after the solution gets opened (for tag helper discovery to not mess with timing here)
- Then I basically, closed and reopened the same file maybe about 20-30 times! 
- (the file is a large razor file (about 900 lines long) zoomed out to 20% to allow request to contain many spans)
- 50-50% it picked out one of the two approaches (to compare before/after side by side)
- Waited about 3 seconds before close and reopening

The screenshots include:
- (a) Average elapsed time for computing semantic using single span vs. multiple spans (before/after)
- (b) Elapsed time for the first 5 requests (before/after) that is triggered when document gets opened in VS

### Key takeaways from the numbers:
- Multiple spans is consistently faster than single span for razor (this PR)
- Nullability analysis when skipped is consistently faster than current default
- The effect with disabling nullability analysis is comparable to (even bigger than) the gain we have when caching is enabled
- The first few calls tend to be slower (I believe each file open triggers three of the requests), mainly the first three seem to be slower.

### CODE REVIEW

**Comment by maryamariyan** [comment]
As far as the PR goes, what I have noted as action item for me is to:
- add to docs mentioning expected input (non-intersecting text spans)
- new tests using multiple text spans

File: src/Workspaces/Core/Portable/Classification/IClassificationService.cs [hunk 1]
```diff
@@ -51,13 +51,13 @@ internal interface IClassificationService : ILanguageService
         /// This will not include classifications for embedded language constructs in string literals.  For that use
         /// <see cref="AddEmbeddedLanguageClassificationsAsync"/>.
         /// </remarks>
-        Task AddSemanticClassificationsAsync(Document document, TextSpan textSpan, ClassificationOptions options, SegmentedList<ClassifiedSpan> result, CancellationToken cancellationToken);
+        Task AddSemanticClassificationsAsync(Document document, ImmutableArray<TextSpan> textSpans, ClassificationOptions options, SegmentedList<ClassifiedSpan> result, CancellationToken cancellationToken);
```
Comment (maryamariyan) [inline src/Workspaces/Core/Portable/Classification/IClassificationService.cs:54]: - [ ] Add to docs mentioning expected input (is non intersecting textspans)

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
```
Comment (ToddGrun) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:8]: nit: are you using this?

File: src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs [hunk 3]
```diff
@@ -56,17 +55,20 @@ internal partial class AbstractSyntaxClassificationService
...
                 CancellationToken cancellationToken)
             {
-                using var worker = new Worker(semanticModel, textSpan, list, getNodeClassifiers, getTokenClassifiers, options, cancellationToken);
+                foreach (var textSpan in textSpans)
+                {
+                    using var worker = new Worker(semanticModel, textSpan, list, getNodeClassifiers, getTokenClassifiers, options, cancellationToken);
 
-                worker._pendingNodes.Push(worker._syntaxTree.GetRoot(cancellationToken));
-                worker.ProcessNodes();
+                    worker._pendingNodes.Push(worker._syntaxTree.GetRoot(cancellationToken));
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:69]: >worker._syntaxTree.GetRoot(cancellationToken) [](http://example.com/codeflow?start=46&length=45)

nit: could push this outside the loop:

var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**